### ISSUE TITLE
Continuous Delivery Pipeline

### ISSUE THREAD

**Issue by prasadtalasila** [prasadtalasila/TransportScheduler#9]
We need to have all code commits regression tested and built. The necessary GitHub integrations must be setup.

**Comment by prasadtalasila** [comment]
I found a high-level release manager for Elixir projects. The [edeliver](https://github.com/boldpoker/edeliver) project wrapping [distillery](https://github.com/bitwalker/distillery) seems like a better option.

Tutorials and video talk are available on [Elixir forum page](https://elixirforum.com/t/elixir-deployment-tools-general-discussion-blog-posts-wiki/827).

Let me see if this integrates with Jenkins. 

How does delivery compare with escript we are currently using?

**Comment by ahaldar** [comment]
The README of edeliver says:
> If using mix, add edeliver and your build tool and your build tool (distillery or exrm) as hex package to your mix.exs config
It seems to be a good option and switching to exrm/distillery (exrm seems to be now deprecated in favor of distillery) should be doable.

**Comment by prasadtalasila** [comment]
exrm replaced by distillery and regression testing to be done before every build.

**Comment by ahaldar** [comment]
To upgrade from exrm to distillery:
https://github.com/boldpoker/edeliver/wiki/Upgrade-from-exrm-to-distillery-as-build-tool

**Comment by ahaldar** [comment]
Travis is the popular option for code repo hosted on GitHub. Travis is a hosted service (free for open source) while you have to host, install and configure Jenkins.
Travis-CI allows a setup for running tests and build release only after success. Our travis.yml will look something like the one here:
https://www.varvet.com/blog/easy-auto-scaling-elixir-deployment-with-aws-and-travis-ci/

**Comment by ahaldar** [comment]
GitHub repo needs to be linked with Travis through owner account.

**Comment by ahaldar** [comment]
Builds work fine on some systems but are showing errors on other systems, gen_statem init function is the point of concern.

**Comment by prasadtalasila** [comment]
This is definitely an issue with OTP v19.1. The existing deployment instructions put in travis as well which is leading to build failure status even for successful builds.

We may need to send an automated ssh password / ssh key for deployment. We may also need a few functional tests on the deployed software.

**Comment by ahaldar** [comment]
Travis builds now passing.
.travis.yml forces versions OTP 19.0 and Elixir 1.3.1

**Comment by ahaldar** [comment]
For Windows deployment, the Docker route seems to be the only option.

**Comment by ahaldar** [comment]
The GenStateMachine README comment doesn't solve the version dependency problem, it only specifies how to enable a certain callback mode that we do not require:
> ...you declare your callback mode as part of use GenStateMachine in this wrapper, rather than returning it from callback_mode/0 on OTP 19.1 and up, or your init/1 and code_change/4 on versions of OTP prior to 19.1. In versions of OTP prior to 19.1, you can still, however, switch callback modes in code_change/4 by returning a callback mode.

Current fix is still .travis.yml specifying OTP 19.0 but Elixir version anything 1.3 and up is fine.

**Comment by prasadtalasila** [comment]
use gen_server + fsm modules for representing a station process.

**Comment by prasadtalasila** [comment]
Not a priority right now.
